contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -77,183 +80,243 @@ contract MixinExchangeWrapper is
-        uint256 zrxBuyAmount,
+        uint256 makerAssetBuyAmount,
         bytes[] memory signatures
     )
         internal
-        returns (FillResults memory totalFillResults)
+        returns (
+            uint256 wethSpentAmount,
+            uint256 makerAssetAcquiredAmount
+        )
     {
-        // Do nothing if zrxBuyAmount == 0
-        if (zrxBuyAmount == 0) {
-            return totalFillResults;
-        }
-
-        bytes memory zrxAssetData = ZRX_ASSET_DATA;
-        bytes memory wethAssetData = WETH_ASSET_DATA;
-        uint256 zrxPurchased = 0;
-
         uint256 ordersLength = orders.length;
         for (uint256 i = 0; i != ordersLength; i++) {
+            if (!orders[i].makerAssetData.equals(orders[0].makerAssetData)) {
+                LibRichErrors.rrevert(LibForwarderRichErrors.MakerAssetMismatchError(
+                    orders[0].makerAssetData,
+                    orders[i].makerAssetData
+                ));
+            }
 
-            // All of these are ZRX/WETH, so we can drop the respective assetData from calldata.
-            orders[i].makerAssetData = zrxAssetData;
-            orders[i].takerAssetData = wethAssetData;
-
-            // Calculate the remaining amount of ZRX to buy.
-            uint256 remainingZrxBuyAmount = _safeSub(zrxBuyAmount, zrxPurchased);
-
-            // Convert the remaining amount of ZRX to buy into remaining amount
-            // of WETH to sell, assuming entire amount can be sold in the current order.
-            // We round up because the exchange rate computed by fillOrder rounds in favor
-            // of the Maker. In this case we want to overestimate the amount of takerAsset.
-            uint256 remainingWethSellAmount = _getPartialAmountCeil(
-                orders[i].takerAssetAmount,
-                _safeSub(orders[i].makerAssetAmount, orders[i].takerFee),  // our exchange rate after fees
-                remainingZrxBuyAmount
-            );
+            // Preemptively skip to avoid division by zero in _marketBuySingleOrder
+            if (orders[i].makerAssetAmount == 0 || orders[i].takerAssetAmount == 0) {
+                continue;
+            }
 
-            // Attempt to sell the remaining amount of WETH.
-            FillResults memory singleFillResult = _fillOrderNoThrow(
+            LibFillResults.FillResults memory singleFillResults = _marketBuySingleOrder(
                 orders[i],
-                remainingWethSellAmount,
-                signatures[i]
+                signatures[i],
+                makerAssetBuyAmount,
+                makerAssetAcquiredAmount
             );
 
-            // Update amounts filled and fees paid by maker and taker.
-            _addFillResults(totalFillResults, singleFillResult);
-            zrxPurchased = _safeSub(totalFillResults.makerAssetFilledAmount, totalFillResults.takerFeePaid);
+            // Percentage fee
+            if (orders[i].takerFeeAssetData.equals(orders[i].makerAssetData)) {
```

Comment:
Similar to above, I think this logic can live in `_marketBuySingleOrder` so that we don't have to repeat the `if...else` blocks.